algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -143,113 +143,150 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+    
+            t_at_beta_s.push(t_at_beta);
+        }
 
+        let verifier::ThirdMessage { r_a, r_b, r_c } = state.third_round_message.as_ref().unwrap();
         let beta = state.second_round_message.unwrap().beta;
         let gamma = state.gamma.unwrap();
 
         let mut linear_combinations = BTreeMap::new();
 
-        let lincheck_time = start_timer!(|| "Lincheck");
         // Lincheck sumcheck:
-        let z_b_s = (0..state.batch_size)
+        let lincheck_time = start_timer!(|| "Lincheck");
+        let z_b_s = BTreeMap::new();
+        for (circuit_hash, circuit_state) in state.circuit_specific_states {
+            let circuit_id = format!("circuit_{:x?}", circuit_hash);
+            let z_b_i = (0..circuit_state.batch_size)
             .map(|i| {
-                let z_b_i = witness_label("z_b", i);
-                LinearCombination::new(z_b_i.clone(), [(F::one(), z_b_i)])
+                let z_b = witness_label(&circuit_id, "z_b", i);
+                LinearCombination::new(z_b.clone(), [(F::one(), z_b)])
             })
             .collect::<Vec<_>>();
+            z_b_s.insert(circuit_hash, &z_b_i);
+        }
+
         let g_1 = LinearCombination::new("g_1", [(F::one(), "g_1")]);
 
         let bivariate_poly_time = start_timer!(|| "Bivariate poly");
-        let r_alpha_at_beta = constraint_domain.eval_unnormalized_bivariate_lagrange_poly(alpha, beta);
+        let r_alpha_at_beta = largest_constraint_domain.eval_unnormalized_bivariate_lagrange_poly(alpha, beta);
         end_timer!(bivariate_poly_time);
 
         let v_H_at_alpha_time = start_timer!(|| "v_H_at_alpha");
-        let v_H_at_alpha = constraint_domain.evaluate_vanishing_polynomial(alpha);
+        let v_H_at_alpha = largest_constraint_domain.evaluate_vanishing_polynomial(alpha);
         end_timer!(v_H_at_alpha_time);
 
         let v_H_at_beta_time = start_timer!(|| "v_H_at_beta");
-        let v_H_at_beta = constraint_domain.evaluate_vanishing_polynomial(beta);
+        let v_H_at_beta = largest_constraint_domain.evaluate_vanishing_polynomial(beta);
         end_timer!(v_H_at_beta_time);
 
         let v_X_at_beta_time = start_timer!(|| "v_X_at_beta");
-        let v_X_at_beta = input_domain.evaluate_vanishing_polynomial(beta);
+        let mut v_X_at_beta = Vec::with_capacity(state.circuit_specific_states.len());
+        for (circuit, circuit_state) in state.circuit_specific_states.iter() {
+            let v_X_i_at_beta = circuit_state.input_domain.evaluate_vanishing_polynomial(beta);
+            v_X_at_beta.push(v_X_i_at_beta);
+        }
         end_timer!(v_X_at_beta_time);
 
-        let z_b_s_at_beta = z_b_s.iter().map(|z_b| evals.get_lc_eval(z_b, beta)).collect::<Result<Vec<_>, _>>()?;
-        let batch_z_b_at_beta: F =
-            z_b_s_at_beta.iter().zip_eq(batch_combiners).map(|(z_b_at_beta, combiner)| *z_b_at_beta * combiner).sum();
+        let z_b_s_at_beta = z_b_s.iter().map(|(circuit, z_b_i)| {
```

Comment:
Let's rename it to `circuit_id`? And maybe give it a new type?

```suggestion
        let z_b_s_at_beta = z_b_s.iter().map(|(circuit, z_b_i)| {
```

Reply:
Done